### ISSUE TITLE
Source Oracle - Memory Leak

### ISSUE THREAD

**Issue by philippeboyd** [airbytehq/airbyte#27846]
### Connector Name

source-oracle

### Connector Version

0.4.0

### What step the error happened?

During the sync

### Revelant information

There seems to be a big memory leak with the `source-oracle` connector. Vertical scaling doesn't help.

I sync about ~1.5 GB of data from difference Oracle sources and the `source-oracle-read` container can take from 9GB up to 18Gb of RAM sometime.

There's no way that syncing 1.5GB of data takes up 18GB of ram in a java process.
<img width="971" alt="Screenshot 2023-06-27 at 10 38 10" src="https://github.com/airbytehq/airbyte/assets/3239656/28196001-c31b-47a1-a64e-2a610b9d3d88">

Might be linked to the other memory leak issue I opened #27844

### Contribute

- [ ] Yes, I want to contribute

**Comment by marcosmarxm** [comment]
@prateekmukhedkar this is something of knowledge of the connectors team?

**Comment by philippeboyd** [comment]
@marcosmarxm @prateekmukhedkar any news on this? We're seeing ram explode for syncing 1GB worth of data
![image](https://github.com/airbytehq/airbyte/assets/3239656/f8663622-0491-40d1-bb41-5462ad00e41f)

**Comment by prateekmukhedkar** [comment]
The amount of memory used will depend on the schema of the tables you have selected for replication. We have also realized that reading all the data from a table in one single read transaction is not the most efficient use of resources, and instead we will be reading data in smaller chunks. We have not scoped out these improvements for source-oracle at present. 

At this point I'll suggest creating a view, limiting data from last X days and/or including certain columns from your source table. In the replication tab, configure a sync for the view you created. Please keep us updated on how things proceed.

**Comment by prateekmukhedkar** [comment]
@philippeboyd I agree with your observation about the memory usage compared to the volume of data synced. At this point I am unable to point to a root cause or whether it is caused by an older dependency. You can either try to upgrade the dependency, recompile the connector and see if it improves the memory usage. If you are willing to do that then I will  provide instructions on how to develop and test connectors locally.

**Comment by philippeboyd** [comment]
Hi @prateekmukhedkar, I indeed tried to update the old OJDBC dependency to latest `21.10.0.0` but I'm facing the same issue. I even tried using OJDBC11; it doesn't change a thing.

**Comment by philippeboyd** [comment]
@prateekmukhedkar After going down the rabbit hole, I would like to have a discussion over [this particular line in the class TwoStageSizeEstimator](https://github.com/airbytehq/airbyte/blob/7a18cc15354f8007c70cec8f784aebb17dd5283e/airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/streaming/TwoStageSizeEstimator.java#L51C9-L51C9)

By doing `getTargetBufferByteSize(Runtime.getRuntime().maxMemory())`, we're essentially telling the connector that all available RAM is open bar for whatever it needs to do...

When I change the code to use `getTargetBufferByteSize(null)` it forces the Estimator's `getTargetBufferByteSize()` to return `MIN_BUFFER_BYTE_SIZE = 250L * 1024L * 1024L; // 250 MB` which helps a LOT with the memory usage without affecting the sync's performance.

Instead of using 15-20GB; it uses 2-3GB.

`git blame` is suggesting me to loop in @tuliren

**Comment by philippeboyd** [comment]
Any news one this? IU'm not even syncing 3GB of data... This memory leak on Source Oracle Read is not normal.

![image](https://github.com/airbytehq/airbyte/assets/3239656/1bd2b355-1a4c-4075-b0f8-5c50c5f7575b)

**Comment by octavia-squidington-iii** [comment]
At Airbyte, we seek to be clear about the project priorities and roadmap. This issue has not had any activity for 365 days, suggesting that it's not as critical as others. It's possible it has already been fixed. It is being marked as stale and will be closed in 20 days if there is no activity. To keep it open, please comment to let us know why it is important to you and if it is still reproducible on recent versions of Airbyte.
